File: client/files2/Área de Trabalho/LP-codes/test/c-python/n-body/n-body.py
import numpy as np

# Parâmetros da simulação
num_bodies = 3
num_steps = 100
dt = 0.1

# Massas e posições iniciais
masses = np.array([10, 5, 3])
positions = np.array([[0, 0], [5, 0], [0, 8]], dtype=np.float64)  # Corrigido o tipo de dados
velocities = np.zeros((num_bodies, 2), dtype=np.float64)  # Corrigido o tipo de dados

def calculate_forces(positions, masses):
    forces = np.zeros_like(positions, dtype=np.float64)
    G = 1.0  # Constante gravitacional

    for i in range(num_bodies):
        for j in range(num_bodies):
            if i != j:
                r = positions[j] - positions[i]
                r_magnitude = np.linalg.norm(r)
                force_magnitude = (G * masses[i] * masses[j]) / (r_magnitude ** 2)
                force = force_magnitude * r / r_magnitude
                forces[i] += force
    
    return forces

def update_positions_and_velocities(positions, velocities, forces, masses, dt):
    accelerations = forces / masses[:, np.newaxis]
    velocities += accelerations * dt
    positions += velocities * dt

# Simulação
for step in range(num_steps):
    print(f"Step {step + 1}:")
    for i in range(num_bodies):
        print(f"Body {i + 1} - Position: {positions[i]}, Velocity: {velocities[i]}")
    
    forces = calculate_forces(positions, masses)
    update_positions_and_velocities(positions, velocities, forces, masses, dt)

print("Simulation completed.")

File: server/backups/files/Área de Trabalho/LP-codes/test/c-python/n-body/n-body.c
#include <stdio.h>
#include <math.h>

#define NUM_BODIES 3
#define NUM_STEPS 100
#define DT 0.1
#define G 1.0

typedef struct {
    double mass;
    double position[2];
    double velocity[2];
} Body;

void calculate_forces(Body bodies[NUM_BODIES], double forces[NUM_BODIES][2]) {
    for (int i = 0; i < NUM_BODIES; i++) {
        forces[i][0] = 0.0;
        forces[i][1] = 0.0;
        
        for (int j = 0; j < NUM_BODIES; j++) {
            if (i != j) {
                double r[2];
                r[0] = bodies[j].position[0] - bodies[i].position[0];
                r[1] = bodies[j].position[1] - bodies[i].position[1];
                double r_magnitude = sqrt(r[0] * r[0] + r[1] * r[1]);
                double force_magnitude = (G * bodies[i].mass * bodies[j].mass) / (r_magnitude * r_magnitude);
                forces[i][0] += force_magnitude * r[0] / r_magnitude;
                forces[i][1] += force_magnitude * r[1] / r_magnitude;
            }
        }
    }
}

void update_positions_and_velocities(Body bodies[NUM_BODIES], double forces[NUM_BODIES][2]) {
    for (int i = 0; i < NUM_BODIES; i++) {
        bodies[i].velocity[0] += forces[i][0] / bodies[i].mass * DT;
        bodies[i].velocity[1] += forces[i][1] / bodies[i].mass * DT;
        
        bodies[i].position[0] += bodies[i].velocity[0] * DT;
        bodies[i].position[1] += bodies[i].velocity[1] * DT;
    }
}

int main() {
    Body bodies[NUM_BODIES] = {
        {10.0, {0.0, 0.0}, {0.0, 0.0}},
        {5.0, {5.0, 0.0}, {0.0, 0.0}},
        {3.0, {0.0, 8.0}, {0.0, 0.0}}
    };

    double forces[NUM_BODIES][2];

    for (int step = 0; step < NUM_STEPS; step++) {
        calculate_forces(bodies, forces);
        update_positions_and_velocities(bodies, forces);
        
        printf("Step %d:\n", step + 1);
        for (int i = 0; i < NUM_BODIES; i++) {
            printf("Body %d: Position (%lf, %lf)\n", i, bodies[i].position[0], bodies[i].position[1]);
        }
        printf("\n");
    }

    return 0;
}
